watertap/examples/flowsheets/electrodialysis/electrodialysis_1stack_conc_recirc.py

```diff
@@ -0,0 +1,460 @@
+    m.fs.costing.add_specific_energy_consumption(
+        m.fs.prod.properties[0].flow_vol_phase["Liq"]
+    )
+
+    # add extra variables and constraints
+    m.fs.recovery_vol_H2O = Var(
+        initialize=0.5,
+        bounds=(0, 1),
+        domain=NonNegativeReals,
+        units=pyunits.dimensionless,
+        doc="flowsheet level water recovery calculated by volumeric flow rate",
+    )
+    m.fs.mem_area = Var(
+        initialize=1,
+        bounds=(0, 1e3),
+        units=pyunits.meter**2,
+        doc="Total membrane area for cem (or aem) in one stack",
+    )
+    m.fs.prod_salinity = Var(
+        initialize=1, bounds=(0, 1000), units=pyunits.kg * pyunits.meter**-3
+    )
+    m.fs.disp_salinity = Var(
+        initialize=1, bounds=(0, 1e6), units=pyunits.kg * pyunits.meter**-3
+    )
+    m.fs.eq_recovery_vol_H2O = Constraint(
+        expr=m.fs.recovery_vol_H2O
+        == m.fs.prod.properties[0].flow_vol_phase["Liq"]
+        * m.fs.feed.properties[0].flow_vol_phase["Liq"] ** -1
+    )
+    m.fs.eq_electrodialysis_equal_flow = Constraint(
+        expr=m.fs.EDstack.diluate.properties[0, 0].flow_vol_phase["Liq"]
+        == m.fs.EDstack.concentrate.properties[0, 0].flow_vol_phase["Liq"]
+    )
+
+    m.fs.eq_product_salinity = Constraint(
+        expr=m.fs.prod_salinity
+        == sum(
+            m.fs.prod.properties[0].conc_mass_phase_comp["Liq", j]
+            for j in m.fs.properties.ion_set | m.fs.properties.solute_set
+        )
+    )
+    m.fs.eq_disposal_salinity = Constraint(
+        expr=m.fs.disp_salinity
+        == sum(
+            m.fs.disp.properties[0].conc_mass_phase_comp["Liq", j]
+            for j in m.fs.properties.ion_set | m.fs.properties.solute_set
+        )
+    )
+
+    m.fs.eq_mem_area = Constraint(
+        expr=m.fs.mem_area
+        == m.fs.EDstack.cell_width
+        * m.fs.EDstack.cell_length
+        * m.fs.EDstack.cell_pair_num
+    )
+
+    # Add Arcs
+    m.fs.arc0 = Arc(source=m.fs.feed.outlet, destination=m.fs.sepa0.inlet)
+    m.fs.arc1b = Arc(source=m.fs.sepa0.to_dil_in, destination=m.fs.pump1.inlet)
+    m.fs.arc1f = Arc(source=m.fs.pump1.outlet, destination=m.fs.EDstack.inlet_diluate)
+    m.fs.arc2 = Arc(
+        source=m.fs.sepa0.to_conc_in0,
+        destination=m.fs.mix0.from_feed,
+    )
+    m.fs.arc3b = Arc(source=m.fs.mix0.outlet, destination=m.fs.pump0.inlet)
+    m.fs.arc3f = Arc(
+        source=m.fs.pump0.outlet, destination=m.fs.EDstack.inlet_concentrate
+    )
+    m.fs.arc4 = Arc(source=m.fs.EDstack.outlet_diluate, destination=m.fs.prod.inlet)
+    m.fs.arc5 = Arc(
+        source=m.fs.EDstack.outlet_concentrate, destination=m.fs.sepa1.inlet
+    )
+    m.fs.arc6 = Arc(source=m.fs.sepa1.to_disp, destination=m.fs.disp.inlet)
+    m.fs.arc7 = Arc(source=m.fs.sepa1.to_conc_in1, destination=m.fs.mix0.from_conc_out)
+    TransformationFactory("network.expand_arcs").apply_to(m)
+
+    return m
+
+
+def condition_base(m):
```

Comment:
Not critical, but several flowsheets (not necessarily all yet), use `set_operating_conditions` as the function name. It might be nice to avoid vastly different naming across flowsheet example.

Reply:
This is a bit different that it leaves two dof to be flexibly controlled.  I need this to efficiently optimize an objective.  I would keep this name to avoid the confusion.

Reply:
Which 2 dof?

Reply:
And why is that needed to optimize? Can you add some more clarification?